### ISSUE TITLE
Table Service: handle sparsity, do not compute cross product between sources and destinations

### ISSUE THREAD

**Issue by tgy** [Project-OSRM/osrm-backend#6466]
I'm facing a bottleneck.

I have a set of pairs `(source, destination)` for which I need to calculate the durations.

Currently I'm passing all my sources, and all my destinations to the [Table Service](http://project-osrm.org/docs/v5.5.1/api/#table-service), which returns a full matrix for all `(source, destination)` pairs.

But this is way too many pairs compared to what I really need.

Is there a way to solve this better with OSRM?

I know it's possible to send one request per pair but I think this is not going to be fast enough for my use.

I saw that in the [General Options](http://project-osrm.org/docs/v5.5.1/api/#general-options) there is a `radiuses` parameter to limit the search, but I'm not sure how to use this with the Table Service.

Any idea or guidance appreciated!

Thanks for this awesome project and any help you could bring!

**Comment by jcoupey** [comment]
You can compute non-square matrices using the `sources` and `destinations` options. Check out the [table service doc](https://github.com/Project-OSRM/osrm-backend/blob/master/docs/http.md#table-service) and examples.

**Comment by tgy** [comment]
True, I can compute non-square matrices. That's actually what I'm already doing. I don't think I said that the number of sources was equal to the number of destinations.

**Comment by jcoupey** [comment]
What about passing all sources and destinations in locations, say $s_1,\ldots,s_n,l_1,\ldots,l_n$ and then set $sources=1;\ldots;n$ and $destinations=n+1;\ldots;2\times n$?

**Comment by tgy** [comment]
This is exactly what I'm already doing :) Which calculates the full $(n, n)$ duration matrix, while I'm only interested in a small sparse part of that matrix.

**Comment by jcoupey** [comment]
> calculates the full duration matrix

You're getting a $n\times n$ matrix while you have $2n$ coordinates in input, which is already not so bad. ;-)

If for all $i$ you need at least one of the $(s_i, l_j)$ values for some $j$ (which you probably do else you'd simply remove $s_i$ from the sources), I don't think you can avoid computing values $(s_i, l_k)$ for the rest of the $k$.

Except if you have huge matrices I would not really bother. Given how OSRM computes the matrices, the overhead of having extra values computed is probably quite low.

Another option would be to split you request in smaller one-to-many requests (or a-few-to-many). For example if you run a request for $s_1$, then you can limit the destinations to the only targets you actually need and not submit all the initial destinations. You could run those in parallel then see whether it's any faster overall.

**Comment by systemed** [comment]
If you literally just want a lot of A–B routes, then running OSRM locally and using either the Node.js bindings, or the code directly as a library in C++, will be super performant. I regularly calculate a batch of over 500,000 routes this way and it breezes through them.

**Comment by tgy** [comment]
Thanks a lot guys for taking the time to answer, this was really helpful.

And sorry for answering only 2 weeks after!

I ended up going for @jcoupey's solution, doing several one-to-many calls to reduce the number of calculated pairs (my problem was sparse and could result in 98% of (orig, dst) pairs being calculated which did not scale (context: ride-hailing app, New Year's Eve high load).

Closing this issue now.

![dist_small_sample_sizes_by_n_pairs](https://user-images.githubusercontent.com/694873/206668351-90064544-b345-4c42-be29-53b750bf0e2c.png)
